services/baseremotecontrol/builtin/builtin.go

```diff
@@ -252,52 +256,117 @@ func (svc *builtIn) ControllerInputs() []input.Control {
+				}
+			}
+			state.mu.Lock()
+			nextLinear, nextAngular = state.linearThrottle, state.angularThrottle
+			state.mu.Unlock()
+
+			if currentLinear != nextLinear || currentAngular != nextAngular {
+				if svc.config.MaxAngularVelocity > 0 && svc.config.MaxLinearVelocity > 0 {
+					if err := svc.base.SetVelocity(
+						svc.cancelCtx,
+						r3.Vector{
+							X: svc.config.MaxLinearVelocity * nextLinear.X,
+							Y: svc.config.MaxLinearVelocity * nextLinear.Y,
+							Z: svc.config.MaxLinearVelocity * nextLinear.Z,
+						},
+						r3.Vector{
+							X: svc.config.MaxAngularVelocity * nextAngular.X,
+							Y: svc.config.MaxAngularVelocity * nextAngular.Y,
+							Z: svc.config.MaxAngularVelocity * nextAngular.Z,
+						},
+						nil,
+					); err != nil {
+						svc.logger.Errorw("error setting velocity", "error", err)
+						if !vutils.SelectContextOrWait(svc.cancelCtx, 10*time.Millisecond) {
+							return
+						}
+						inRetry = true
+						continue
+					}
+				} else {
+					if err := svc.base.SetPower(svc.cancelCtx, nextLinear, nextAngular, nil); err != nil {
+						svc.logger.Errorw("error setting power", "error", err)
+						if !vutils.SelectContextOrWait(svc.cancelCtx, 10*time.Millisecond) {
+							return
+						}
+						inRetry = true
+						continue
+					}
+				}
+				inRetry = false
+
+				currentLinear = nextLinear
+				currentAngular = nextAngular
+			}
+		}
+	}, svc.activeBackgroundWorkers.Done)
+}
+
 func (svc *builtIn) processEvent(ctx context.Context, state *throttleState, event input.Event) {
-	newLinear, newAngular := parseEvent(svc.controlMode, state, event)
 
+	// Order of who processes what event is *not* guranteed. It depends on the mutex
+	// fairness mode. Ordering logic must be handled at a higher level in the robot.
+	// Other than that, values overwrite each other.
 	state.mu.Lock()
-	if similar(newLinear, state.linearThrottle, .05) && similar(newAngular, state.angularThrottle, .05) {
-		state.mu.Unlock()
-		return
+	oldLinear := state.linearThrottle
+	oldAngular := state.angularThrottle
+	newLinear := oldLinear
+	newAngular := oldAngular
+
+	switch svc.controlMode {
+	case joyStickControl:
+		newLinear.Y, newAngular.Z = oneJoyStickEvent(event, state.linearThrottle.Y, state.angularThrottle.Z)
+	case droneControl:
+		newLinear, newAngular = droneEvent(event, state.linearThrottle, state.angularThrottle)
+	case triggerSpeedControl:
+		newLinear.Y, newAngular.Z = triggerSpeedEvent(event, state.linearThrottle.Y, state.angularThrottle.Z)
+	case buttonControl:
+		newLinear.Y, newAngular.Z, state.buttons = buttonControlEvent(event, state.buttons)
+	case arrowControl:
+		newLinear.Y, newAngular.Z, state.arrows = arrowEvent(event, state.arrows)
 	}
+	state.linearThrottle = newLinear
+	state.angularThrottle = newAngular
 	state.mu.Unlock()
 
-	session.SafetyMonitor(ctx, svc.base)
+	if similar(newLinear, oldLinear, .05) && similar(newAngular, oldAngular, .05) {
+		return
+	}
 
-	svc.activeBackgroundWorkers.Add(1)
```

Comment:
instead of piling on a ton of goroutines, we just track our goal throttle state while allowing any commutative properties from control mode + parsed event.